triples-generator/src/main/scala/ch/datascience/triplesgenerator/reprovisioning/OutdatedTriplesFinder.scala

```diff
@@ -41,85 +40,120 @@ private class IOOutdatedTriplesFinder(
     extends IORdfStoreClient[RdfQuery](rdfStoreConfig, logger)
     with OutdatedTriplesFinder[IO] {
 
-  override def findOutdatedTriples: OptionT[IO, OutdatedTriples] = OptionT {
-    queryExpecting[Option[OutdatedTriples]](using = query)
+  import io.circe.Decoder
+  import io.circe.Decoder._
+
+  override def findOutdatedTriples: OptionT[IO, OutdatedTriples] =
+    for {
+      projectPath     <- maybeProjectWithOutdatedAgent orElse maybeProjectWithNoAgent
+      outdatedTriples <- findOutdatedTriples(projectPath)
+    } yield outdatedTriples
+
+  private def maybeProjectWithOutdatedAgent = OptionT {
+    queryExpecting[Option[FullProjectPath]] {
+      s"""
+         |PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+         |PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+         |PREFIX prov: <http://www.w3.org/ns/prov#>
+         |PREFIX schema: <http://schema.org/>
+         |PREFIX dcterms: <http://purl.org/dc/terms/>
+         |
+         |SELECT ?project
```

Comment:
could you do `SELECT DISTINCT` here without a `LIMIT` at the end of the query? You would then get the full list of projects you need to process right away.

Reply:
Actually, that's intended to pick up a single project with outdated triples. The way, the re-provisioning process works is that it deals with a single project at a time to prevent the 'stop the world' situations. Even more, the process picks up a new project only if there are no interactive requests being served at the moment.

Reply:
aha ok - I guess if these queries turn out to be lightweight it should be fine. I was just thinking to be more economical with the volume of requests. However, won't it take a long time to reprovision all the projects on renku version change this way?

Reply:
It could take a bit longer but the idea is that it's a bit more lightweight and more 'backgroundy' so e.g. when users do pushes to GitLab, these get served first.